### PR TITLE
Properly track DLC transactions in wallet

### PR DESCRIPTION
**Pull request by benthecarman**
Necessary for upcoming GUI work. This will have the wallet use the spk dao to track and look for the funding and closing txs

### CODE REVIEW

File: dlc-wallet/src/main/scala/org/bitcoins/dlc/wallet/DLCWallet.scala [hunk 1]
```diff
@@ -106,6 +106,38 @@ abstract class DLCWallet extends Wallet with AnyDLCHDWalletApi {
     } yield updated
   }
 
+  private def updateFundingOutPoint(
```
Comment (Christewart) [inline dlc-wallet/src/main/scala/org/bitcoins/dlc/wallet/DLCWallet.scala:110]: nit: Can you add a scaladoc? What is it updating? It looks like it updates `DLCDb` adding the outpoint to the database?

File: dlc-wallet/src/main/scala/org/bitcoins/dlc/wallet/DLCWallet.scala [hunk 2]
```diff
@@ -106,6 +106,38 @@ abstract class DLCWallet extends Wallet with AnyDLCHDWalletApi {
...
+      contractId: ByteVector,
+      outPoint: TransactionOutPoint): Future[DLCDb] = {
```
Comment (Christewart) [inline dlc-wallet/src/main/scala/org/bitcoins/dlc/wallet/DLCWallet.scala:112]: nit: Hmm... I guess I don't have a strong preference here, but do you think it's better to throw an exception rather than just returning `Future[Option[DLCDb]]`? Same applies with the `updateClosingTxId` method.

File: dlc-wallet/src/main/scala/org/bitcoins/dlc/wallet/DLCWallet.scala [hunk 3]
```diff
@@ -521,7 +582,15 @@ abstract class DLCWallet extends Wallet with AnyDLCHDWalletApi {
             case (input, tx) => input.toFundingInput(tx.transaction)
           })
 
-          updatedDLCDb <- updateDLCContractIds(offer, accept)
+          dlcDb <- updateDLCContractIds(offer, accept)
+
+          builder = DLCTxBuilder(offer, accept.withoutSigs)
+          fundingTx <- builder.buildFundingTx
+          outPoint = TransactionOutPoint(fundingTx.txId, UInt32.zero)
```
Comment (Christewart) [inline dlc-wallet/src/main/scala/org/bitcoins/dlc/wallet/DLCWallet.scala:655]: just curious, why is it safe to use outpiont index `0`?

File: dlc-wallet/src/main/scala/org/bitcoins/dlc/wallet/DLCWallet.scala [hunk 4]
```diff
@@ -890,7 +969,10 @@ abstract class DLCWallet extends Wallet with AnyDLCHDWalletApi {
 
     executorFromDb(dlcDb, dlcOffer, dlcAccept, fundingInputs)
       .flatMap { executor =>
-        val outcomeSigs = outcomeSigDbs.map(_.toTuple).toMap
+        val outcomeSigs = outcomeSigDbs
+          .filter(_.isInitiator == !dlcDb.isInitiator)
```
Comment (Christewart) [inline dlc-wallet/src/main/scala/org/bitcoins/dlc/wallet/DLCWallet.scala:1040]: This deserves a comment, what is the plain english explaination of what's going on here?

**Review by Christewart — CHANGES_REQUESTED** [review]
It doesn't look like there are any test cases added to make sure we actually update these transactions through the flow of a DLC. It seems to me we need 

1. A test case that makes sure we update the funding txid 
2. A test case that makes sure we update the closing txid where we close the DLC
3. A test that makes sure we update the closing txid in the case where our remote peer closes the DLC
4. A test that makes sure we update the closing txid for the refund case

File: dlc-wallet/src/main/scala/org/bitcoins/dlc/wallet/DLCWallet.scala (hunk 3 above)
Comment (benthecarman) [reply dlc-wallet/src/main/scala/org/bitcoins/dlc/wallet/DLCWallet.scala:655]: Currently, the funding output is always at that index

**Comment by benthecarman** [comment]
> It doesn't look like there are any test cases added to make sure we actually update these transactions through the flow of a DLC. It seems to me we need
> 
>     1. A test case that makes sure we update the funding txid
> 
>     2. A test case that makes sure we update the closing txid where we close the DLC
> 
>     3. A test that makes sure we update the closing txid in the case where our remote peer closes the DLC
> 
>     4. A test that makes sure we update the closing txid for the refund case

Those are all tested with `verifyProperlySetTxIds`

File: dlc-wallet/src/main/scala/org/bitcoins/dlc/wallet/DLCWallet.scala (hunk 2 above)
Comment (benthecarman) [reply dlc-wallet/src/main/scala/org/bitcoins/dlc/wallet/DLCWallet.scala:112]: I think this is better, we are only going to be calling this function inside of other functions that should have already verified that the `contractId` exists. Making this an option will just make things unnecessarily complicated

Comment (Christewart) [reply dlc-wallet/src/main/scala/org/bitcoins/dlc/wallet/DLCWallet.scala:112]: Ok fair enough, it sees we just return a `T` everywhere else rather than an `Option[T]` when we use the term `update`

**Comment by Christewart** [comment]
> > It doesn't look like there are any test cases added to make sure we actually update these transactions through the flow of a DLC. It seems to me we need
> > ```
> > 1. A test case that makes sure we update the funding txid
> > 
> > 2. A test case that makes sure we update the closing txid where we close the DLC
> > 
> > 3. A test that makes sure we update the closing txid in the case where our remote peer closes the DLC
> > 
> > 4. A test that makes sure we update the closing txid for the refund case
> > ```
> 
> Those are all tested with `verifyProperlySetTxIds`

Ahh ok!

File: dlc-wallet/src/main/scala/org/bitcoins/dlc/wallet/DLCWallet.scala [hunk 5]
```diff
@@ -151,6 +185,83 @@ abstract class DLCWallet extends Wallet with AnyDLCHDWalletApi {
...
+    super.processOutgoingUtxos(transaction, blockHashOpt).flatMap { res =>
+      val outPoints = transaction.inputs.map(_.previousOutput).toVector
+
+      for {
+        dlcDbs <- dlcDAO.findByFundingOutPoints(outPoints)
+        _ <-
+          if (dlcDbs.nonEmpty) {
+            val txDb = TransactionDbHelper.fromTransaction(transaction)
+            transactionDAO.upsert(txDb)
+          } else FutureUtil.unit
```
Comment (nkohen) [inline dlc-wallet/src/main/scala/org/bitcoins/dlc/wallet/DLCWallet.scala]: nit: Probably deserves its own util

File: dlc-wallet/src/main/scala/org/bitcoins/dlc/wallet/DLCWallet.scala [hunk 6]
```diff
@@ -988,9 +1139,11 @@ abstract class DLCWallet extends Wallet with AnyDLCHDWalletApi {
   override def executeDLCRefund(contractId: ByteVector): Future[Transaction] = {
     for {
       (executor, setup) <- executorAndSetupFromDb(contractId)
+      outcome = executor.executeRefundDLC(setup)
+      _ <- processTransaction(outcome.refundTx, None)
```
Comment (nkohen) [inline dlc-wallet/src/main/scala/org/bitcoins/dlc/wallet/DLCWallet.scala]: nit: Name None

File: wallet/src/main/scala/org/bitcoins/wallet/Wallet.scala [hunk 7]
```diff
@@ -205,6 +205,7 @@ abstract class Wallet
   override def broadcastTransaction(transaction: Transaction): Future[Unit] =
     for {
       _ <- nodeApi.broadcastTransaction(transaction)
+      _ <- processTransaction(transaction, None)
```
Comment (nkohen) [inline wallet/src/main/scala/org/bitcoins/wallet/Wallet.scala]: nit: name None

**Review by nkohen — APPROVED** [review]
Looking good

File: dlc-wallet/src/main/scala/org/bitcoins/dlc/wallet/DLCWallet.scala [hunk 8]
```diff
@@ -151,6 +185,83 @@ abstract class DLCWallet extends Wallet with AnyDLCHDWalletApi {
...
+                DLCState.Refunded
+              } else if (dlcDb.state == DLCState.Claimed) {
+                DLCState.Claimed
+              } else {
+                DLCState.RemoteClaimed
+              }
+
+            dlcDb.copy(state = state)
+        }
+      case (_, _) => Future.successful(dlcDb)
```
Comment (Christewart) [inline dlc-wallet/src/main/scala/org/bitcoins/dlc/wallet/DLCWallet.scala]: Please don't use `case  (_,_) =>`, enumerate the cases so we retain pattern match exhaustiveness 

```scala
case (Some(_),None) | (None, Some(_)) | (None,None) => Future.successful(dlcDb)
```

File: dlc-wallet/src/main/scala/org/bitcoins/dlc/wallet/DLCWallet.scala [hunk 9]
```diff
@@ -151,6 +185,83 @@ abstract class DLCWallet extends Wallet with AnyDLCHDWalletApi {
...
+    }
+  }
+
+  override def processIncomingUtxos(
```
Comment (Christewart) [inline dlc-wallet/src/main/scala/org/bitcoins/dlc/wallet/DLCWallet.scala]: scaladoc on this please, it seems 

```
/** Process an incoming utxo as normal, and then update the DLC state for the incoming utxo */ 
```

Is that right?

File: dlc-wallet/src/main/scala/org/bitcoins/dlc/wallet/DLCWallet.scala [hunk 10]
```diff
@@ -151,6 +185,83 @@ abstract class DLCWallet extends Wallet with AnyDLCHDWalletApi {
...
+            transactionDAO.upsert(txDb)
+          } else FutureUtil.unit
+
+        // Update the state to be confirmed or broadcasted
+        updated = dlcDbs.map { dlcDb =>
+          dlcDb.state match {
+            case DLCState.Offered | DLCState.Accepted | DLCState.Signed |
+                DLCState.Broadcasted =>
+              if (blockHashOpt.isDefined)
+                dlcDb.copy(state = DLCState.Confirmed)
```
Comment (Christewart) [inline dlc-wallet/src/main/scala/org/bitcoins/dlc/wallet/DLCWallet.scala:229]: Isn't this circumventing our normal confirmation policy settings? Why isn't this the same as a the normal transaction confirmation policy we have?

Comment (Christewart) [reply dlc-wallet/src/main/scala/org/bitcoins/dlc/wallet/DLCWallet.scala:229]: Also, whenever we get around to fixing #2179 we would also need to update this code to handle the reorg properly right?

It seems we should really try and centralize the "confirmation" logic ine one place rather than one offs.

File: wallet/src/main/scala/org/bitcoins/wallet/internal/TransactionProcessing.scala [hunk 11]
```diff
@@ -156,7 +156,7 @@ private[wallet] trait TransactionProcessing extends WalletLogger {
       }
     }
 
-  private def processIncomingUtxos(
+  def processIncomingUtxos(
```
Comment (Christewart) [inline wallet/src/main/scala/org/bitcoins/wallet/internal/TransactionProcessing.scala]: can this be `protected`?

Comment (Christewart) [reply wallet/src/main/scala/org/bitcoins/wallet/internal/TransactionProcessing.scala]: but as per my suggestion earlier in this PR, i think this should be refactored into a utility method that also takes in a `WalletApi` as a parameter so something like this


```scala
def processIncomingUtxos(
      transaction: Transaction,
      blockHashOpt: Option[DoubleSha256DigestBE],
      newTags: Vector[AddressTag],
      wallet: WalletApi): Future[Vector[SpendingInfoDb]] = ...
```

That way we can call it inside of `DLCWallet` as well? Does that work?

File: wallet/src/main/scala/org/bitcoins/wallet/internal/TransactionProcessing.scala [hunk 12]
```diff
@@ -438,18 +438,26 @@ private[wallet] trait TransactionProcessing extends WalletLogger {
 
         for {
           _ <- insertIncomingTransaction(transaction, totalIncoming)
+
+          addrs <- addressDAO.findAllAddresses()
```
Comment (Christewart) [inline wallet/src/main/scala/org/bitcoins/wallet/internal/TransactionProcessing.scala:445]: Hmm, why is this needed now?

**Review by Christewart — CHANGES_REQUESTED** [review]

File: dlc-wallet/src/main/scala/org/bitcoins/dlc/wallet/DLCWallet.scala (hunk 10 above)
Comment (benthecarman) [reply dlc-wallet/src/main/scala/org/bitcoins/dlc/wallet/DLCWallet.scala:229]: This isn't the TxoState, this is the DLCState.

File: wallet/src/main/scala/org/bitcoins/wallet/internal/TransactionProcessing.scala (hunk 12 above)
Comment (benthecarman) [reply wallet/src/main/scala/org/bitcoins/wallet/internal/TransactionProcessing.scala:445]: To differentiate the spks that belong to the wallet and the ones that don't. For example, we don't want to add the DLC funding output to our spending info db as it will effect our balance, and we could not spend it.

File: wallet/src/main/scala/org/bitcoins/wallet/internal/TransactionProcessing.scala (hunk 11 above)
Comment (benthecarman) [reply wallet/src/main/scala/org/bitcoins/wallet/internal/TransactionProcessing.scala]: We can't do this, this is an implementation specific function. If we used a `WalletApi` then we wouldn't have access to our DAOs

File: dlc-wallet/src/main/scala/org/bitcoins/dlc/wallet/DLCWallet.scala (hunk 10 above)
Comment (Christewart) [reply dlc-wallet/src/main/scala/org/bitcoins/dlc/wallet/DLCWallet.scala:229]: They are inherently linked. What happens in the case of reorg? Or is `DLCState.Confirmed` just named badly? It seems you are setting this based off of a blockhash, so it does have some relationship to actually being confirmed in the chain.

Comment (benthecarman) [reply dlc-wallet/src/main/scala/org/bitcoins/dlc/wallet/DLCWallet.scala:229]: Oh yes, in that context it has the same problem. Was more so responding to your first comment

Comment (Christewart) [reply dlc-wallet/src/main/scala/org/bitcoins/dlc/wallet/DLCWallet.scala:229]: Hmm, so does this expand the scope of #2179 then? Or is it the same problem?

**Review by Christewart — APPROVED** [review]
